io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -435,6 +438,20 @@ data TraceEvent
   | EventTimerExpired   TimeoutId
   deriving Show
 
+-- | Get a trace of 'EventSay'.
+--
+traceSayEvents :: Trace a -> [String]
+traceSayEvents = go []
+    where
+      go acc (Trace _  _ (EventSay s) tr) = go (s : acc) tr
+      go acc (Trace _  _ _            tr) = go acc tr
+      go acc _                            = reverse acc
```

Comment:
This should not be tail recursive with an accumulator. It should return results incrementally.

I would model it on this version, and indeed we should include this one too:
```
-- | Select all the traced values matching the expected type. This relies on
-- the sim's dynamic trace facility.
--
-- For convenience, this throws exceptions for abnormal sim termination.
--
selectTraceDynamic :: Typeable b => Trace a -> [b]
selectTraceDynamic = go
  where
    go (Trace _ _ (EventLog dyn) trace) = case fromDynamic dyn of
                                            Just x  -> x : go trace
                                            Nothing ->     go trace
    go (Trace _ _ _              trace) = go trace
    go (TraceMainException _ e _)       = throw (FailureException e)
    go (TraceDeadlock      _   _)       = throw FailureDeadlock
    go (TraceMainReturn    _ _ _)       = []
```

Note that it returns results incrementally and also throws an error at the end on sim failures which is usually what one wants in such situations.

Reply:
Good point, I'll fix this.